cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java

```diff
@@ -382,41 +394,174 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
+        List<InputField> inputs = write.getInputs();
+        for (InputField input : inputs) {
+          if (input.getName().equals(destinationField.getField())) {
+            // if the write operation did write the destinationField
+            // mark this write operation as visited
+            incomingOperations.add(write);
+            // traverse backward in the graph by looking up the origin of this input field which is the operation
+            // which computed this destinationField
+            computeIncoming(new EndPointField(write.getDestination(), input.getName()),
+                            operationsMap.get(input.getOrigin()), write, incomingOperations,
+                            new HashMap<>());
+          }
+        }
       }
+    }
+    return incomingOperations;
+  }
+
+  /**
+   * <p>Get the subset of operations that used the specified field</p>
+   *
+   * <p>For example if the operation are as follow</p>
+   * <pre>
+   * pRead: personFile -> (offset, body)
+   * parse: body -> (id, name, address)
+   * cRead: codeFile -> id
+   * codeGen: (parse.id, cRead.id) -> id
+   * sWrite: (codeGen.id, parse.name, parse.address) -> secureStore
+   * iWrite: (parse.id, parse.name, parse.address) -> insecureStore
+   * </pre>
+   *
+   * <p>The sourceField  is 'id' of codeFile then the returned operations set will contain cRead, codeGen, sWrite.</p>
+   * <p>If the sourceField is 'body' of personFile then the returned set of operations will contain pRead, parse,
+   * codeGen, sWrite, iWrite.</p>
+   *
+   * @param sourceField the {@link EndPointField} whose outgoing operations needs to be found
+   *
+   * @return {@link Set} of {@link Operation} which are outgoing from the given sourceField
+   */
+  Set<Operation> getOutgoingOperationsForField(EndPointField sourceField) {
+    if (readOperations == null) {
+      computeAndValidateFieldLineageInfo(this.operations);
+    }
+
+    Set<Operation> outgoingOperations = new HashSet<>();
+    for (ReadOperation readOperation : readOperations) {
+      if (readOperation.getSource().equals(sourceField.getEndPoint())) {
+        // if the read operation is for the dataset to which the sourceField belong
+        List<String> inputs = readOperation.getOutputs();
+        for (String input : inputs) {
+          if (input.equals(sourceField.getField())) {
+            // if the read operation read the sourceField
+            // add the read operation to outgoingOperations
+            outgoingOperations.add(readOperation);
+            // find all the outgoing operation from this read operations and if they are connected i.e. the next
+            // operation did used the field for which the outgoing operations is requested then traverse that
+            // operation further
+            for (Operation outgoingOperation : operationOutgoingConnections.get(readOperation.getName())) {
```

Comment:
this much nesting makes things difficult to understand. I think it would be cleaner to do something like:

```
ReadOperation readOperation = null;
for (ReadOperation operation : readOperations) {
  if (operation.getSource().equals(sourceField.getEndPoint()) {
    operation = readOperation;
    break;
  }
}

boolean fieldIsOutput = false;
for (String output : readOperation.getOutputs()) {
  if (output.equals(sourceField.getField())) {
    fieldIsOutput = true;
    break;
  }
}

if (fieldIsOutput) {
  ...
}
```

This would also prevent unnecessary looping that is currently happening.